python/ray/data/datasource/avro_datasource.py

```diff
@@ -0,0 +1,53 @@
+from io import BytesIO
+from typing import TYPE_CHECKING, Iterator, List, Union
+
+from ray.data._internal.delegating_block_builder import DelegatingBlockBuilder
+from ray.data.block import Block
+from ray.data.datasource.file_based_datasource import FileBasedDatasource
+from ray.util.annotations import PublicAPI
+
+if TYPE_CHECKING:
+    import pyarrow
+
+
+@PublicAPI(stability="alpha")
+class AvroDatasource(FileBasedDatasource):
+    """Avro datasource, for reading and writing Avro files."""
+
+    _FILE_EXTENSIONS = ["avro"]
+
+    def __init__(
+        self,
+        paths: Union[str, List[str]],
+        **file_based_datasource_kwargs,
+    ):
+        super().__init__(paths, **file_based_datasource_kwargs)
+
+    def _read_stream(self, f: "pyarrow.NativeFile", path: str) -> Iterator[Block]:
+        import fastavro
+
+        builder = DelegatingBlockBuilder()
+
+        # Check if the file is empty
+        # Assuming `f.readall()` reads the entire Avro file into memory
+        file_content = f.readall()
+        if not file_content:
+            yield builder.build()  # Yield an empty block for an empty file
+            return
```

Comment:
What happens if we don't do this? With this, we're reading each file twice

Reply:
This empty file test will fail: https://github.com/ray-project/ray/pull/43663/files#diff-8c3ba0c18b56d9d6425ef8cc8c63b5a4bbf1d32c375e420b4b1c5f3afcb64aa4R44-R48

Since the reader will fail due to empty file. I can have a try on empty file with schema.

Reply:
removed, thanks